### PR TITLE
`Development`: Add LTI Authentication for email address in use

### PR DESCRIPTION
**Pull request by basak-akan**
> [!WARNING]  
> **Deploy only locally or on ts11!** The LTI feature is only available on TS11.
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://ls1intum.github.io/Artemis/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://ls1intum.github.io/Artemis/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/client/).

### Motivation and Context

We identified a potential security issue where an instructor could potentially log in as another user when manipulating LTI auth data and email address. 

### Description

When a user attempts to access Artemis and isn't already logged in on the same browser, the Artemis server identifies this and sends a 401 error to the client, indicating that authentication is required. This error is then captured on the client side for appropriate handling. Conversely, if the user is already logged into Artemis on the same browser, the client recognizes this authenticated state and automatically redirects the user to the appropriate page without further prompts.

### Steps for Testing

Prerequisites:
- 1 Student

User not logged into Artemis:
1. Log out from Artemis if you are already logged in with a user.
2. Log in to https://moodle.ase.in.tum.de/ using credentials:
  -  username: student1 
  - password: same as artemis_test_user_1
3. Navigate to "TS11 Authentication Test Exercise" in "LTI Testing Course" or simply go to https://moodle.ase.in.tum.de/mod/lti/view.php?id=127
4. Verify you are forwarded to login page
5. Login Artemis using artemis_test_user_1 credentials
6. Verify you are directed into the text exercise

User logged into Artemis:
1. Login to Artemis using artemis_test_user_1 credentials
2. Log in to https://moodle.ase.in.tum.de/ using credentials:
  -  username: student1 
  - password: same as artemis_test_user_1
3. Navigate to "TS11 Authentication Test Exercise" in "LTI Testing Course" or simply go to https://moodle.ase.in.tum.de/mod/lti/view.php?id=127
4. Verify you are directed into the text exercise

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [lti13-exercise-launch.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5520/latest/artifact/shared/Coverage-Report-Client-Tests/app/lti/lti13-exercise-launch.component.ts.html) | 100% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [LtiEmailAlreadyInUseException.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5520/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.exception/LtiEmailAlreadyInUseException.html) | 100% | ✅ |
| [Lti13LaunchFilter.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5520/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.security.lti/Lti13LaunchFilter.html) | 93% | ✅ |
| [Lti13Service.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5520/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.lti/Lti13Service.html) | 95% | ✅ |
| [LtiService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5520/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.lti/LtiService.html) | 97% | ✅ |

### Screenshots

User is already logged in to Artemis in another browser tab:
LTI_Auth

User is not logged in to Artemis:
LTI_AUTH_2

### CODE REVIEW

**Review by jakubriegel — DISMISSED** [review]
Works, demonstrated on a testing session

**Review by Kroko-fant — APPROVED** [review]
Tested in a testing session and worked as expected!

**Review by rstief — APPROVED** [review]
Shown in testing session, looks good.

File: src/main/java/de/tum/in/www1/artemis/security/lti/Lti13LaunchFilter.java [hunk 1]
```diff
@@ -53,11 +54,14 @@ protected void doFilterInternal(HttpServletRequest request, HttpServletResponse
             filterChain.doFilter(request, response);
             return;
         }
-
+        String targetLink = "";
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/security/lti/Lti13LaunchFilter.java:57]: It took me a second to understand why this has to be initialized here / why the catch block sets this possibly empty header in the field. Maybe you can add a comment for future devs?

File: src/main/webapp/app/lti/lti13-exercise-launch.component.ts [hunk 2]
```diff
@@ -57,9 +64,30 @@ export class Lti13ExerciseLaunchComponent implements OnInit {
                         console.error('No LTI targetLinkUri received for a successful launch');
                     }
                 },
-                error: () => {
```
Comment (MaximilianAnzinger) [inline src/main/webapp/app/lti/lti13-exercise-launch.component.ts:60]: This is quite lengthy. Could you separate the next and error handling in methods to make this a little bit easier to understand?

**Review by MaximilianAnzinger — CHANGES_REQUESTED** [review]
Other than my two optional code style suggestions the code looks good 👍

**Review by lennart-keller — DISMISSED** [review]
Besides the (optional) suggestions from Max, code lgtm :)

File: src/main/java/de/tum/in/www1/artemis/security/lti/Lti13LaunchFilter.java (hunk 1 above)
Comment (basak-akan) [reply src/main/java/de/tum/in/www1/artemis/security/lti/Lti13LaunchFilter.java:59]: Thank you for your suggestion! I added some comments for targetLink initialization and its usage in LtiEmailAlreadyInUseException catch block to explain it better.

File: src/main/webapp/app/lti/lti13-exercise-launch.component.ts (hunk 2 above)
Comment (basak-akan) [reply src/main/webapp/app/lti/lti13-exercise-launch.component.ts:60]: I've separated the next and error handling into distinct methods, hope it is more understandable now.

**Review by MaximilianAnzinger — DISMISSED** [review]
Thank you for considering my suggestions. Code LGTM 👍

**Comment by xHadie** [comment]
Manual test on TS11 worked just like you described 👍

**Comment by JanaNF** [comment]
Manual tested on ts11.

1. User not logged into Artemis: Worked successfully
2. User logged into Artemis: Worked successfully

**Review by lennart-keller — APPROVED** [review]
Code lgtm now :)

**Review by Kroko-fant — DISMISSED** [review]
Looks good to me👍

**Review by jakubriegel — APPROVED** [review]
Code update ok

**Review by rstief — APPROVED** [review]

**Review by JanaNF — DISMISSED** [review]
Manual tested on ts11.

User not logged into Artemis: Worked successfully
User logged into Artemis: Worked successfully

**Review by jakubriegel — APPROVED** [review]
Thanks for additional tests

**Review by lennart-keller — APPROVED** [review]
Reapprove code after adding tests.

**Review by MaximilianAnzinger — APPROVED** [review]
Code

**Pull request merged**